Add support for INDEX BY an associated entity (2.8.x)

andrews05 opened the pull request:
This allows specifying an association in the INDEX BY clause of a query
which will index by the association's join column.

Related to #7661.

beberlei commented:
@andrews05 awesome :-) Can you fix the coding standard violatoins listed here? 

https://travis-ci.org/doctrine/orm/jobs/616171121?utm_medium=notification&utm_source=github_status

This should fix automatically when you run `php vendor/bin/phpcbf lib/Doctrine/ORM/Query/SqlWalker.php` and commit the changes.

andrews05 commented:
I did run phpcs but it seems the <3.0 branches don't really adhere to the coding standard? It showed 301 errors in that file alone and phpcbf "failed to fix". I could fix the issues in my changes manually, if you think it's worth it? (Keeping in my mind that code is copied from elsewhere in the file)

andrews05 committed:
Add support for INDEX BY an associated entity

This allows specifying an association in the INDEX BY clause of a query
which will index by the association's join column.

Related to #7661.

greg0ire commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -854,9 +856,41 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
      */
     public function walkIndexBy($indexBy)
     {
-        $pathExpression = $indexBy->simpleStateFieldPathExpression;
```
Why did you rename this variable?

andrews05 replied:
To answer all three of your comments, the switch statement here was simply copied from walkPathExpression(). I renamed this variable only because it was easier than changing it throughout the switch.

greg0ire commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -854,9 +856,41 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
+                break;
+
+            case AST\PathExpression::TYPE_SINGLE_VALUED_ASSOCIATION:
+                // 1- the owning side:
```
Why put a number if that comment?

greg0ire replied:
Found the commit that introduced it: 639718e95c1ee483d0c5f889b0326ec39aa187c5

greg0ire commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -854,9 +856,41 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
+                    $class = $this->em->getClassMetadata($class->associationMappings[$fieldName]['inherited']);
+                }
+
+                $assoc = $class->associationMappings[$fieldName];
```
Please refrain from using abbrevations, `association` is fine.

greg0ire reviewed (commented):
I have a few nitpicks, but this looks great 👍

Should this be documented somewhere?

beberlei commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -1613,7 +1613,7 @@ From, Join and Index by
     RangeVariableDeclaration                   ::= AbstractSchemaName ["AS"] AliasIdentificationVariable
     JoinAssociationDeclaration                 ::= JoinAssociationPathExpression ["AS"] AliasIdentificationVariable [IndexBy]
     Join                                       ::= ["LEFT" ["OUTER"] | "INNER"] "JOIN" (JoinAssociationDeclaration | RangeVariableDeclaration) ["WITH" ConditionalExpression]
-    IndexBy                                    ::= "INDEX" "BY" StateFieldPathExpression
```
Please also add an example in the docs below after the language description in the `Using INDEX BY ` section.

andrews05 replied:
I've added a second example. Does it look okay?
I think the existing example is helpful for a more complex result with nested arrays and multiple levels of indexing, so I left that one there.

andrews05 commented:
@lcobucci Any chance of this getting approved?

lcobucci commented on lib/Doctrine/ORM/Query/AST/IndexBy.php:
```diff
@@ -33,14 +33,14 @@ class IndexBy extends Node
      * @var PathExpression
      */
-    public $simpleStateFieldPathExpression = null;
+    public $singleValuedPathExpression = null;
```
Since this class isn't marked as `@internal` renaming the property is a BC-break

lcobucci commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -854,9 +856,40 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
         $alias          = $pathExpression->identificationVariable;
-        $field          = $pathExpression->field;
+
+        /** @var Query\AST\PathExpression $pathExpression */
```
We prefer `assert()` over `/** @var */`

andrews05 replied:
Sorry, I wasn't entirely clear of the purpose of some of the docblocks like this. Perhaps the `@var` is redundant as the variable isn't a function parameter?

greg0ire replied:
If you don't need it for auto-complete or SA tools like PHPStan or Psalm, it should be removed, otherwise please use `assert($pathExpression instanceof PathExpression)` (and import that class of course). It will have the same effect but might be checked in some setups, which makes it better (less likely to rot than a comment).

lcobucci commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -854,9 +856,40 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
-        $field          = $pathExpression->field;
+
+        /** @var Query\AST\PathExpression $pathExpression */
+        switch ($pathExpression->type) {
+            case AST\PathExpression::TYPE_STATE_FIELD:
+                $field = $pathExpression->field;
+                break;
+
+            case AST\PathExpression::TYPE_SINGLE_VALUED_ASSOCIATION:
+                // Just use the foreign key, i.e. u.group_id
+                $fieldName = $pathExpression->field;
+                $class     = $this->queryComponents[$alias]['metadata'];
+
+                if (isset($class->associationMappings[$fieldName]['inherited'])) {
+                    $class = $this->em->getClassMetadata($class->associationMappings[$fieldName]['inherited']);
+                }
+
+                $association = $class->associationMappings[$fieldName];
+
+                if (! $association['isOwningSide']) {
+                    throw QueryException::associationPathInverseSideNotSupported($pathExpression);
+                }
+
+                if (count($association['sourceToTargetKeyColumns']) > 1) {
+                    throw QueryException::associationPathCompositeKeyNotSupported();
+                }
+
+                $field = reset($association['targetToSourceKeyColumns']);
+                break;
+
+            default:
+                throw QueryException::invalidPathExpression($pathExpression);
+        }
```
This seems like something that could be extracted to a private method and possibly use `if` and early exits (allowing for us to have a more strict comparison).

lcobucci replied:
Also, is this done elsewhere?

andrews05 replied:
This is largely a copy of the switch from walkPathExpression(), though not identical. The walkPathExpression() one is building a sql string wheras this one is just determining the appropriate field.
Should I try combining them both into a private method?

lcobucci commented:
Would you please rebase this to resolve the conflicts? I'm slightly concerned with the BC-break but I trust @beberlei and @greg0ire to make a decision here and ship this :+1:

andrews05 commented:
@lcobucci Conflicts resolved. Should I squash the commits also?

greg0ire commented:
> Should I squash the commits also?

Yes IMO

andrews05 commented:
@greg0ire Commits squashed. I removed the @var comment, it's redundant for autocomplete.

beberlei committed:
Reintroduce IndexBy#simpleStateFieldPathExpression as deprecated property.

beberlei committed:
Merge 2.9.x into andrews05-indexby

beberlei committed:
Housekeeping: phpcs

beberlei committed:
Housekeeping: phpcs
